packages/contracts/contracts/L2/L2StandardBridge.sol

```diff
@@ -0,0 +1,186 @@
+            // uses onlyEOA on the receive function, so anyone sending to a contract knows
+            // what they are doing.
+            finalizeBridgeETH(_from, _to, _amount, _data);
+            emit DepositFinalized(_l1Token, _l2Token, _from, _to, _amount, _data);
+        } else if (
+            _isOptimismMintable(_l2Token, _l1Token)
+        ) // Check the target token is compliant and
+        // verify the deposited token on L1 matches the L2 deposited token representation here
+        // slither-disable-next-line reentrancy-events
+        {
+            // When a deposit is finalized, we credit the account on L2 with the same amount of
+            // tokens.
+            // slither-disable-next-line reentrancy-events
+            finalizeBridgeERC20(_l2Token, _l1Token, _from, _to, _amount, _data);
+            // slither-disable-next-line reentrancy-events
+            emit DepositFinalized(_l1Token, _l2Token, _from, _to, _amount, _data);
+        } else {
+            // Either the L2 token which is being deposited-into disagrees about the correct address
+            // of its L1 token, or does not support the correct interface.
+            // This should only happen if there is a  malicious L2 token, or if a user somehow
+            // specified the wrong L2 token address to deposit into.
+            // In either case, we stop the process here and construct a withdrawal
+            // message so that users can get their funds out in some cases.
+            // There is no way to prevent malicious token contracts altogether, but this does limit
+            // user error and mitigate some forms of malicious contract behavior.
+            emit DepositFailed(_l1Token, _l2Token, _from, _to, _amount, _data);
+
+            // Withdraw ETH in the case that the user submitted a bad ETH
+            // deposit to prevent ETH from getting stuck
+            // TODO: can this be wrapped into _initiateWithdrawal?
+            // need to handle using `msg.value` here instead of `_value`
+            if (_l2Token == Lib_PredeployAddresses.OVM_ETH) {
+                _initiateBridgeETH(_from, _to, msg.value, 0, _data);
+            } else {
+                _initiateBridgeERC20(_l2Token, _l1Token, _from, _to, _amount, 0, _data);
+            }
+        }
+    }
+
+    /**********************
+     * Internal Functions *
+     **********************/
+
+    /**
+     * @notice Handle withdrawals, taking into account the legacy form of ETH
+     * when it was represented as an ERC20 at the OVM_ETH contract.
+     * TODO: require(msg.value == _value) for OVM_ETH case?
+     */
+    function _initiateWithdrawal(
+        address _l2Token,
+        address _from,
+        address _to,
+        uint256 _amount,
+        uint32 _minGasLimit,
+        bytes calldata _data
+    ) internal {
+        address l1Token = OptimismMintableERC20(_l2Token).l1Token();
```

Comment:
Could use `remoteToken()` to avoid legacy interface.

Reply:
Needs to be l1Token() or will break legacy contracts